### DISCUSSION TITLE
Client Server with ESP32

### DISCUSSION THREAD

**Discussion by A622266** [micropython/micropython#9714] [Hardware & Peripherals] [answered]
I am trying to generate a signal on the ESP32 (the client) and to send it to a python script on a host PC (the server).  On my home network, the host PC is on 192.168.1.22.  The protocol is UDP, and the server listens on port 56000 for any data.  The following code implements the server running on the PC:

```py
#  This is the server in CPython
import socket
port = 56000
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.bind(("", port))
print(f"waiting on port: {port}")
while 1:
	data, addr = s.recvfrom(1024)
	print(data)
```
The client connects to the network on address 192.168.1.11, and sends data using port 56000.  ```esp_network``` is a function to connect the esp32 to the network.  ```make_tone``` is a function derived from the ```play_tone``` example from Mike Teachman's [i2s-examples](https://github.com/miketeachman/micropython-i2s-examples/tree/master/examples).  The following code implements the client running on the esp32:
```py
# this is the client in micropython
import usocket as socket
import esp_network
import make_tone

esp_network.do_connect()

# ======= AUDIO CONFIGURATION =======
TONE_FREQUENCY_IN_HZ = 5000
SAMPLE_SIZE_IN_BITS = 16
SAMPLE_RATE_IN_HZ = 48_000
# ======= AUDIO CONFIGURATION =======

ADDR = "192.168.1.11"      # This is IP of the server
PORT = 56000

# Socket to exchange data with GR
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

while True:
    samples = make_tone.make_tone(SAMPLE_RATE_IN_HZ, 
        SAMPLE_SIZE_IN_BITS, TONE_FREQUENCY_IN_HZ)
    print(samples)
    s.sendto(samples, (ADDR, PORT))
```
The esp32 successfully connects to the network on 192.168.1.11.  ```samples``` is successfully generated as a bytearray, and generally the client appears to work.  However, the server does not receive the samples.  I tried changing ```ADDR``` to be ```192.168.1.22```, but I get an error:  ```OSError: [Errno 12] ENOMEM```. 

Can anyone see my error, or make a suggestion of something to try?

**Comment by davefes** [comment]
Try printing the addr in the server while loop to confirm.

I do something similar but send a message from the server, then at the client get the server's address.  Then you can print it out to confirm that you have a working link.

**Reply by davefes** [reply]
Where does esp_network come from?

**Reply by A622266** [reply]
```esp_network``` is derived from the function found in the quick-doc for esp32:
```py
def do_connect():
    import network
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    if not wlan.isconnected():
        print('connecting to network...')
        wlan.connect('ssid', 'key')
        while not wlan.isconnected():
            pass
    print('network config:', wlan.ifconfig())
```
I tried printing ```addr``` from the server code as you suggested.  But it doesn't execute, I assume  because no data is ever received from the client.

**Reply by davefes** [reply]
The do_connect code is familiar.  I am guessing you have put it in a file called esp_network.py.

Get the ESP32 running and try to ping it from the PC to determine if the ESP32 is even connected to your network.  Until you get things running and maybe even later is to have a blinking LED in your client loop just so you know it is running or monitor its operation with something like rshell.

**Reply by A622266** [reply]
Yes I put that code in a file called esp_network.py
However, I tried your idea to ping from a PC -- and I get 100% packet loss.  Interestingly, if I go to my router control panel, my esp32 appears to be on the network on 192.168.1.11, which is why I thought everything was fine.  I tried ping some other websites, and other devices on my network, and those work.  So I guess my esp32 is not really connected.  Any further thoughts?

**Reply by A622266** [reply]
I made some progress -- I had the esp32 on a guest network.  The PC where the server is located is on a different network, using the same router.  I moved the esp32 to the same network as the PC, and now I can ping it without packet loss.  
Still I cannot get data to move from the client to the server.  Is my code correct that ```ADDR=192.168.1.11```, as that is the address assigned by my router to the esp32?  As noted above, I get an error if I change this to ```ADDR=192.168.1.22```, which is the address that my PC is on.

**Reply by davefes** [reply]
```py
print('network config:', wlan.ifconfig())
```
should have told you if it was connected and to who.  What does it say?

**Reply by davefes** [reply]
> As noted above, I get an error if I change this to ADDR=192.168.1.22, which is the address that my PC is on.

That should sent it to the PC, so something else is wrong.

**Reply by davefes** [reply]
There some suggestions way down this posting:
[https://forum.pycom.io/topic/2210/errno-12-enomem/4](url)

**Reply by davefes** [reply]
I noticed a comment about UDP needing two-way comms.  That is what I do but don't know it is really required.

Also, when you finished sending data you need to close the socket ... if you finish!  I am just sending commands to an ESP32, so I do not stay connected.

**Reply by A622266** [reply]
This is the output when ```ADDR=192.168.1.22```:
```
network config: ('192.168.1.11', '255.255.255.0', '192.168.1.1', '192.168.1.1')
bytearray(b'\x00\x04\x91\x06\xef\x07u\x07]\x05\xa3\x02\x8b\x00\x11\x00o\x01')
bytearray(b'\x00\x04\x91\x06\xef\x07u\x07]\x05\xa3\x02\x8b\x00\x11\x00o\x01')
bytearray(b'\x00\x04\x91\x06\xef\x07u\x07]\x05\xa3\x02\x8b\x00\x11\x00o\x01')
bytearray(b'\x00\x04\x91\x06\xef\x07u\x07]\x05\xa3\x02\x8b\x00\x11\x00o\x01')
Traceback (most recent call last):
  File "<stdin>", line 23, in <module>
OSError: [Errno 12] ENOMEM
```
It seems to go through the loop 4 times before crashing.
If ```ADDR=192.168.1.11```, the loop runs continuously.  In the case, the network config line is the same as above.

I can ping ```192.168.1.11``` and ```192.168.1.22``` successfully.

**Reply by A622266** [reply]
Okay I'll look into the matter of 2 way comms for UDP -- I don't think it should be required, but better to verify.   Can you check the link?  It does not appear to work..

**Reply by davefes** [reply]
Hmmm ... is:
```py
s.recvfrom(1024)
```
too small?  Try increasing it and when that fails you will probably need to do some reading on how to deal with continuous data.  I haven't done that yet.

**Reply by davefes** [reply]
[https://stackoverflow.com/questions/56951343/send-continuous-data-in-python-with-udp-socket](url)
See last comment.

**Reply by davefes** [reply]
I always send a reply back to the server to ACK correct reception.  Sounds like a step towards TCP!

**Reply by davefes** [reply]
Ctrl + K and Ctrl + C, according to the GitHub docs
[https://stackoverflow.com/questions/56951343/send-continuous-data-in-python-with-udp-socket](url)
OK I give up.  How does one add a link?

**Comment by davefes** [comment]
Looks like I don't how to add links correctly!  Copy/paste and see if that works.

**Reply by A622266** [reply]
on the links:  you use the little icon 6th from the left that looks like a chain.  You highlight a word like [Wikipedia](https://www.wikipedia.org/), and then past the url into the highlighted section that shows up after you click the icon.
I think you are onto something.  If I make the server buffer bigger, I do get some data transfers before crashing!  Probably I need a correct buffer size and a correct way of handling continuous data..

**Reply by davefes** [reply]
That is what I was doing in the first place :)  Ah, when you highlight the link icon it actually says Ctrl K, so I didn't need to go searching the docs.
Maybe, it is a Ubuntu thing.

**Comment by A622266** [comment, marked as answer]
Okay I think I have a path now, thanks to help from davefes.  The server buffer size matters to make sure you capture all the data sent, but it doesn't appear to cause the crash.  The crash is caused by not having a delay in the ```while``` loop at client -- so I suppose if the data sent is more than the esp32 wifi can keep up with, there is a crash.  If I add a delay as follows:
```
while True:
    samples = make_tone.make_tone(SAMPLE_RATE_IN_HZ, 
        SAMPLE_SIZE_IN_BITS, TONE_FREQUENCY_IN_HZ)
    print(samples)
    s.sendto(samples, (ADDR, PORT))
    time.sleep(0.1)
```
There is no crash, and the server receives the data.  I'm not sure I can do what I want to do, which is to send audio data in real time over the wifi.  If the data is 16-bit, at 48khz sample rate, that's 768kpbs worst case.  Can anyone comment on if this is too fast?

**Reply by davefes** [reply]
I would suggest marking this answered and start a new thread.  Also, I get many of my socket-related answers on stackoverflow.

**Reply by A622266** [reply]
agreed, thanks for your help!
